libraries/eosiolib/contracts/eosio/key_value_singleton.hpp

```diff
@@ -0,0 +1,164 @@
+            memcpy(ste.raw_original.data(), val_buffer, copy_size);
+
+            if (value_size > detail::max_stack_buffer_size) {
+               free(val_buffer);
+            }
+         }
+
+         return get_state().value;
+      }
+
+      void set(const T& val) {
+         auto& ste = get_state();
+         ste.value = val;
+         ste.is_dirty = true;
+         ste.is_cached = true;
+      }
+
+      void erase() {
+         internal_use_do_not_use::kv_erase(db_name, contract_name.value, key.data(), key.size());
+         auto& ste = get_state();
+         ste.is_cached = false;
+         ste.is_dirty = false;
+      }
+
+      void store() {
+         auto& ste = get_state();
+         if (ste.is_dirty) {
+            size_t data_size = get_size(ste.value);
+            void* data_buffer = data_size > detail::max_stack_buffer_size ? malloc(data_size) : alloca(data_size);
+
+            serialize(ste.value, data_buffer, data_size);
+
+            if (ste.raw_original.size() != data_size || memcmp(ste.raw_original.data(), data_buffer, data_size) != 0) {
+               internal_use_do_not_use::kv_set(db_name, contract_name.value, key.data(), key.size(), (const char*)data_buffer, data_size);
+            }
+         }
+      }
+
+   private:
+      constexpr static uint64_t db_name = static_cast<uint64_t>(DbName);
+      constexpr static uint64_t singleton_name = static_cast<uint64_t>(SingletonName);
+
+      eosio::name contract_name;
+      key_type key;
+
+      key_type make_prefix(uint64_t singleton_name, uint8_t status = 2) {
+         return make_key(std::make_tuple(status, singleton_name));
+      }
+
+      template <typename V>
+      static void serialize(const V& value, void* buffer, size_t size) {
+         datastream<char*> ds((char*)buffer, size);
+         unsigned_int i{0};
+         ds << i;
+         ds << value;
+      }
+
+      template <typename... Vs>
+      static void serialize(const std::variant<Vs...>& value, void* buffer, size_t size) {
+         datastream<char*> ds((char*)buffer, size);
+         ds << value;
+      }
+
+      template <typename V>
+      static void deserialize(V& value, void* buffer, size_t size) {
+         unsigned_int idx;
+         datastream<const char*> ds((char*)buffer, size);
+
+         ds >> idx;
+         eosio::check(idx==unsigned_int(0), "there was an error deserializing this value.");
+         ds >> value;
+      }
+
+      template <typename... Vs>
+      static void deserialize(std::variant<Vs...>& value, void* buffer, size_t size) {
+         datastream<const char*> ds((char*)buffer, size);
+         ds >> value;
+      }
+
+      template <typename V>
+      static size_t get_size(const V& value) {
+         auto size = pack_size(value);
+         return size + 1;
+      }
+
+      template <typename... Vs>
+      static size_t get_size(const std::variant<Vs...>& value) {
+         auto size = pack_size(value);
+         return size;
+      }
```

Comment:
This duplicates code in key_value.hpp

Reply:
Yeah I'm planning on moving this to the `key_value.hpp` file once all the other comments are resolved (I'm not sure GH would track that move correctly).